File: python/gunmu/decorators.py
"""@gunmu 装饰器。

把返回 None 的函数自动包成 Nothing，返回正常值自动包成 Some。
也支持直接返回 Gunmu 的函数（原样透传）。
"""
from __future__ import annotations

import functools
import inspect
from typing import Any, Callable

from .core import Gunmu, Nothing, Some


def gunmu(
    reason: str | None = None,
) -> Callable[[Callable[..., Any]], Callable[..., Gunmu[Any]]]:
    """装饰器工厂。

    用法：
        @gunmu()
        def find(uid): ...

        @gunmu(reason="没查到")
        def find(uid): ...
    """

    def decorator(func: Callable[..., Any]) -> Callable[..., Gunmu[Any]]:
        default_reason = reason or f"{func.__name__} 返回了 None"

        if inspect.iscoroutinefunction(func):

            @functools.wraps(func)
            async def async_wrapper(*args: Any, **kwargs: Any) -> Gunmu[Any]:
                result = await func(*args, **kwargs)
                return _wrap(result, default_reason)

            return async_wrapper  # type: ignore[return-value]

        @functools.wraps(func)
        def wrapper(*args: Any, **kwargs: Any) -> Gunmu[Any]:
            result = func(*args, **kwargs)
            return _wrap(result, default_reason)

        return wrapper

    return decorator


def _wrap(result: Any, reason: str) -> Gunmu[Any]:
    if isinstance(result, Gunmu):
        return result
    if result is None:
        return Nothing(reason)
    return Some(result)

File: typescript/gunmu/decorators.ts
/**
 * @gunmu 装饰器。
 *
 * 把返回 null / undefined 的函数自动包成 Nothing，
 * 返回正常值自动包成 Some。
 */

import { Gunmu, some, nothing } from "./core.js";

type AnyFn = (...args: any[]) => any;

export function gunmu(reason?: string) {
  return function decorator<F extends AnyFn>(
    fn: F,
  ): (...args: Parameters<F>) => Promise<Gunmu<Awaited<ReturnType<F>>>> {
    const defaultReason = reason ?? `${fn.name || "anonymous"} 返回了 null`;

    return async function wrapped(
      ...args: Parameters<F>
    ): Promise<Gunmu<Awaited<ReturnType<F>>>> {
      const result = await fn(...args);
      return wrap(result, defaultReason);
    };
  };
}

function wrap<T>(result: T, reason: string): Gunmu<T> {
  if (
    result !== null &&
    typeof result === "object" &&
    "kind" in result &&
    ((result as any).kind === "some" || (result as any).kind === "nothing")
  ) {
    return result as unknown as Gunmu<T>;
  }
  if (result === null || result === undefined) {
    return nothing(reason);
  }
  return some(result);
}